src/Ocelot/Multiplexer/MultiplexingMiddleware.cs

```diff
@@ -1,219 +1,265 @@
+    /// <returns>The cloned context.</returns>
+    private static HttpContext CreateThreadContext(HttpContext source)
+    {
+        var target = new DefaultHttpContext
         {
-            if (route.DownstreamRoute.Count > 1)
+            Request =
             {
-                var aggregator = _factory.Get(route);
-                await aggregator.Aggregate(route, httpContext, contexts);
-            }
-            else
+                Body = source.Request.Body,// Consider stream cloning for multiple reads
+                ContentLength = source.Request.ContentLength,
+                ContentType = source.Request.ContentType,
+                Host = source.Request.Host,
+                Method = source.Request.Method,
+                Path = source.Request.Path,
+                PathBase = source.Request.PathBase,
+                Protocol = source.Request.Protocol,
+                QueryString = source.Request.QueryString,
+                Scheme = source.Request.Scheme,
+                IsHttps = source.Request.IsHttps,
+                Query = new QueryCollection(new Dictionary<string, StringValues>(source.Request.Query)),
+                RouteValues = new RouteValueDictionary(source.Request.RouteValues),
+            },
+            Connection =
             {
-                // Assume at least one... if this errors then it will be caught by global exception handler
-                MapNotAggregate(httpContext, contexts.First());
-            }
-        }
-
-        private static void MapNotAggregate(HttpContext httpContext, HttpContext finished)
+                RemoteIpAddress = source.Connection.RemoteIpAddress,
+            },
+            RequestServices = source.RequestServices,
+            RequestAborted = source.RequestAborted,
+            User = source.User,
+        };
+
+        foreach (var header in source.Request.Headers)
         {
-            httpContext.Items.UpsertErrors(finished.Items.Errors());
-
-            httpContext.Items.UpsertDownstreamRequest(finished.Items.DownstreamRequest());
-
-            httpContext.Items.UpsertDownstreamResponse(finished.Items.DownstreamResponse());
+            target.Request.Headers[header.Key] = header.Value.ToArray();
         }
 
-        private static async Task<HttpContext> Fire(HttpContext httpContext, RequestDelegate next)
+        return target;
+    }
+
+    protected virtual Task Map(HttpContext httpContext, Route route, List<HttpContext> contexts)
+    {
+        if (route.DownstreamRoute.Count == 1)
         {
-            await next.Invoke(httpContext);
-            return httpContext;
+            return Task.CompletedTask;
         }
+
+        var aggregator = _factory.Get(route);
+        return aggregator.Aggregate(route, httpContext, contexts);
+    }
+
+    private static async Task<HttpContext> Fire(HttpContext httpContext, RequestDelegate next)
+    {
+        await next.Invoke(httpContext);
+        return httpContext;
```

Comment:
Not your code, but I cannot get it, why do we have to await here?
And, why do we return a `HttpContext` result instead of `void`?

Reply:
Yes your right we could return Task, we are not reusing the httpContext.

Reply:
And we will have single line returning expression which is simple wrapping sentence doing only a call.
So, `Fire` method is boring wrapper , not helper. And it can be removed.
My concern when I reviewed the middleware for the first time I was confused with `Fire` name.
To fire what? Event? Propagate object?
But events are propagated by `Raise` method in .NET. 
My opinion: we need to remove this method.

Reply:
Agree!

Reply:
@raman-m we can remove it, but, but. we need to wait on the next, because the http context will be modified. Because, under the next, it's the response coming back.